src/base/intrusive_list.h

```diff
@@ -48,26 +48,56 @@ namespace perfetto::base {
 
 namespace internal {
 
+// ListNode is used both in (1) actual list nodes; (2) as the list head, to
+// point to the first and last element in the list.
+// When prev_/next_ of a node point to the list head, the address has the LSB
+// set. This has two purposes: (1) identify the list head to stop the iterator;
+// (2) prevent bugs which try to dereference the list head casting it into a T.
 struct ListNode {
-  ListNode* prev;
-  ListNode* next;
+  uintptr_t prev_;
+  uintptr_t next_;
 };
 
+// This function masks away the LSB retutning a pointer to a ListNode. This can
+// be used when we want to dereference a prev_/next_ pointer and we acknowledge
+// that we might be operating on the list head rather than a node.
+inline ListNode* MaybeHead(uintptr_t p) {
+  return reinterpret_cast<ListNode*>(p & ~uintptr_t(1));
+}
+
 // IntrusiveList's Base class to factor out type-independent code (avoid binary
 // bloat)
 class ListOps {
  public:
   void PushFront(ListNode* node);
+  void PushBack(ListNode* node);
   void PopFront();
+  void PopBack();
   void Erase(ListNode* node);
+  bool empty() const { return head_.next_ == head_masked(); }
 
-  ListNode* front_{nullptr};
-  size_t size_{0};
+  uintptr_t head_masked() const {
```

Comment:
Comment on this or naming something else would be useful. Is the most mysterious feeling part of this CL, not sure it's the best name you could use.

Reply:
Yeah `head_with_sentinel` or `head_and_sentinel` seems good.